contracts/pool/src/tests/swap.rs

```diff
@@ -1027,8 +1061,198 @@ fn test_should_fail_when_invalid_ask_asset_min_amount() {
+        None,
+        stake_manager,
+        stake_owner,
+    );
+
+    token1.mint(&user1, &1_001_000);
+    token2.mint(&user1, &1_001_000);
+
+    env.ledger().with_mut(|li| li.timestamp = 49);
+    pool.provide_liquidity(
+        &user1,
+        &Some(1_000_000),
+        &Some(1_000_000),
+        &Some(1_000_000),
+        &Some(1_000_000),
+        &None,
+        &Some(50u64),
+    );
+
+    // selling just one token with 1% max spread allowed
+    let spread = 100i64; // 1% maximum spread allowed
+    env.ledger().with_mut(|li| li.timestamp = 99);
+    // we set the deadline to be at latest 100 and we execute swap at 99
+    pool.swap(
+        &user1,
+        // FIXM: Disable Referral struct
+        // &None::<Referral>,
+        &token1.address,
+        &1,
+        &None,
+        &Some(spread),
+        &Some(100u64),
+    );
+
+    let share_token_address = pool.query_share_token_address();
+    let result = pool.query_pool_info();
+    assert_eq!(
+        result,
+        PoolResponse {
+            asset_a: Asset {
+                address: token1.address.clone(),
+                amount: 1_000_001i128,
+            },
+            asset_b: Asset {
+                address: token2.address.clone(),
+                amount: 999_999i128,
+            },
+            asset_lp_share: Asset {
+                address: share_token_address.clone(),
+                amount: 1_000_000i128,
+            },
+            stake_address: result.clone().stake_address,
+        }
+    );
+    assert_eq!(token1.balance(&user1), 999); // -1 from the swap
+    assert_eq!(token2.balance(&user1), 1001); // 1 from the swap
+
+    // this time 100 units
+    env.ledger().with_mut(|li| li.timestamp = 149);
+    let output_amount = pool.swap(
+        &user1,
+        // FIXM: Disable Referral struct
+        // &None::<Referral>,
+        &token2.address,
+        &1_000,
+        &None,
+        &Some(spread),
+        &Some(150),
+    );
+    let result = pool.query_pool_info();
+    assert_eq!(
+        result,
+        PoolResponse {
+            asset_a: Asset {
+                address: token1.address.clone(),
+                amount: 1_000_001 - 1_000, // previous balance minus 1_000
+            },
+            asset_b: Asset {
+                address: token2.address.clone(),
+                amount: 999_999 + 1000,
+            },
+            asset_lp_share: Asset {
+                address: share_token_address,
+                amount: 1_000_000i128, // this has not changed
+            },
+            stake_address: result.clone().stake_address,
+        }
+    );
+    assert_eq!(output_amount, 1000);
+    assert_eq!(token1.balance(&user1), 1999); // 999 + 1_000 as a result of swap
+    assert_eq!(token2.balance(&user1), 1001 - 1000); // user1 sold 1k of token B on second swap
+}
+
+#[test]
+#[should_panic(expected = "Pool: Swap: Transaction executed after deadline!")]
+fn simple_swap_with_should_fail_when_after_the_deadline() {
```

Comment:
This is fine, but you need  to cover providing and withdrawing liquidity as well.